pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,252 @@
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutInterface string
+	InInterface  string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Interface interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	InitMultiClusterMssClamp() error
+	DelMultiClusterMssClamp() error
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	UnInstallGlobalEgress() error
+}
+
+type Driver interface {
+	Interface
+	MssClampInterface
+	GlobalNetEgressInterface
+}
+type DriverCreateFunc func() (Driver, error)
+
+var (
+	logger         = log.Logger{Logger: logf.Log.WithName("PacketFilter")}
+	drivers        = map[string]DriverCreateFunc{}
+	selectedDriver string
+)
+
+func SetDriver(name string) {
+	logger.Infof("driver set to: %q", name)
+	selectedDriver = name
+}
+
+func GetDriver() string {
+	return selectedDriver
+}
+
+func AddDriver(name string, createFunc DriverCreateFunc) {
+	if drivers[name] != nil {
+		logger.Warningf("Multiple drivers attempting to register with name %q", name)
```

Comment:
I think we should treat this error as FATAL or return an appropriate error here.

Reply:
I think we should do away with adding drivers as I mentioned here https://github.com/submariner-io/submariner/pull/2835#discussion_r1432722928 and simply let the caller figure out which driver to use and pass it to `New`, `func New(d Driver) (Interface, error)`. Unit tests would pass in the fake `Driver`.